### PR TITLE
[RLlib] Change (prioritized) SA episode buffer to return episode lists (instead of batch) from `sample()`.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Originally from @sven1977.

This PR proposes off-policy sampling in form of episode chunk lists to unify the `Learner API` for SA/MA settings. Until data arrives in the learner connectors it is in episode format - only in the learner we create batches. 

This will be the first of a series of PRs implementing this for all algorithms and settings (MA/SA). 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py [hunk 1]
```diff
@@ -309,6 +305,159 @@ def sample(
...
+            B += 1
+
+            # Keep track of sampled indices for updating priorities later.
+            self._last_sampled_indices.append(idx)
+
+        self.sampled_timesteps += batch_size_B
+
+        return sampled_episodes
+
+    # @override(EpisodeReplayBuffer)
```
Comment (sven1977) [inline rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py:449]: nit: should we remove `OLD_sample` and `sample_by_keys` then?

Would any argument speak for a flexible buffer that can return both episodes AND batches?
I tend to argue no, but not 100% sure yet.

File: rllib/utils/replay_buffers/utils.py [hunk 2]
```diff
@@ -18,39 +18,36 @@
...
                 continue
 
-            from ray.rllib.algorithms.dqn.dqn_rainbow_learner import TD_ERROR_KEY
-
-            # Get the TD-error from the results.
-            td_error = result_dict.get(TD_ERROR_KEY, None)
-
             # Warn once, if we have no TD-errors to update priorities.
-            if td_error is None:
+            if "td_error" not in td_error or td_error["td_error"] is None:
```
Comment (sven1977) [inline rllib/utils/replay_buffers/utils.py:50]: super nit: Can we use the `TD_ERROR_KEY` constant here?

**Review by sven1977 — APPROVED** [review]
Awesome rescuing of my other broken PR! Thanks @simonsays1980 for fixing the root problem with the correct terminateds/truncateds in the returned episodes!

LGTM now.

2 nits/questions before we can merge.

File: rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/utils/replay_buffers/prioritized_episode_replay_buffer.py]: I say no, too - and have removed it

**Comment by simonsays1980** [comment]
@sven1977 have to ping you here b/c my anyscale account was deleted (see my email). Changes include your proposals from the review. Docs give an error that appears to be rooted elsewhere

**Pull request merged**